### PR TITLE
PT-1382 Reduce password reset key length for patients

### PR DESCRIPTION
**Pull request by franck-fourel**
No change for guests
Patients will receive a 8 characters code that will have to be submitted against a new route `confirm/accept/patient/forgot`

### CODE REVIEW

File: api/forgot.go [hunk 1]
```diff
@@ -69,8 +69,8 @@ func (a *Api) passwordReset(res http.ResponseWriter, req *http.Request, vars map
 			log.Print(STATUS_RESET_PATIENT)
 			log.Printf("email used [%s]", email)
 			resetCnf, _ = models.NewConfirmation(models.TypePatientPasswordReset, models.TemplateNamePatientPasswordReset, "")
-			// a patient is not allowed to reset his password, close the request
-			resetCnf.UpdateStatus(models.StatusCompleted)
+			// // a patient is not allowed to reset his password, close the request
+			// resetCnf.UpdateStatus(models.StatusCompleted)
```
Comment (ydrisR) [inline api/forgot.go]: Shouldn't we remove these comments?

File: templates/locales/fr.yaml [hunk 2]
```diff
@@ -26,6 +26,7 @@ PatientPasswordResetHello: "Bonjour,"
 PatientPasswordResetRequest: "Vous avez demandé une réinitialisation de votre mot de passe YourLoops."
 PatientPasswordResetDidNotRequest: "Si vous n'êtes pas à l'origine de cette demande, merci d'ignorer cet email."
 PatientPasswordResetByDevice: "La réinitialisation de votre mot de passe s'effectue via votre terminal Diabeloop."
+PatientPasswordResetExplainKey: "Suivez les instructions en utilisant la clé ci-dessous sur votre terminal Diabeloop."
```
Comment (ydrisR) [inline templates/locales/fr.yaml]: This should not be updated here. Only in Crowdin

File: api/hydrophoneApi.go [hunk 3]
```diff
@@ -133,6 +133,7 @@ func (a *Api) SetHandlers(prefix string, rtr *mux.Router) {
 	accept := rtr.PathPrefix("/accept").Subrouter()
 	accept.Handle("/signup/{confirmationid}", varsHandler(a.acceptSignUp)).Methods("PUT")
 	accept.Handle("/forgot", varsHandler(a.acceptPassword)).Methods("PUT")
+	accept.Handle("/patient/forgot", varsHandler(a.acceptPassword)).Methods("PUT")
```
Comment (ydrisR) [inline api/hydrophoneApi.go]: Why add a new route since it calls the same method at the end?

File: api/forgot_test.go [hunk 4]
```diff
@@ -38,6 +38,26 @@ func TestForgotResponds(t *testing.T) {
 				"password": "myN3wpa55w0rd",
 			},
 		},
+		{
```
Comment (ydrisR) [inline api/forgot_test.go:55]: Need to add tests to verify the new behaviour of method `passwordReset` when requester is a patient

File: clients/mockStoreClient.go [hunk 5]
```diff
@@ -50,8 +50,12 @@ func (d *MockStoreClient) FindConfirmation(notification *models.Confirmation) (r
 	if notification.Email == "email.resend@address.org" {
 		notification.TemplateName = models.TemplateNameSignup
 	}
-
-	notification.Created = time.Now().AddDate(0, 0, -3) // created three days ago
+	if notification.ShortKey == "12345678" {
+		onehour, _ := time.ParseDuration("1h")
+		notification.Created = time.Now().Add(onehour) // created three days ago
```
Comment (ydrisR) [inline clients/mockStoreClient.go]: created one hour ago?

File: models/confirmation_test.go [hunk 6]
```diff
@@ -123,3 +187,17 @@ func TestConfirmationKey(t *testing.T) {
...
+	keyLength := 8
+	key := generateShortKey(keyLength)
+
+	if key == "" {
+		t.Fatal("There should be a generated key")
+	}
+
+	if len(key) != keyLength {
+		t.Fatal("The generated key should be 8 chars: ", len(key))
+	}
```
Comment (ydrisR) [inline models/confirmation_test.go:214]: Perhaps also verify the key only contains alpha/numeric characters?

**Review by ydrisR — CHANGES_REQUESTED** [review]
It works well (tested on the docker stack) 👍 
Somme comments to review

File: models/confirmation.go [hunk 7]
```diff
@@ -69,19 +71,26 @@ const (
...
 
 var (
 	Timeouts TypeDurations = TypeDurations{
-		TypeCareteamInvite: 7 * 24 * time.Hour,
-		TypePasswordReset:  7 * 24 * time.Hour,
-		TypeSignUp:         31 * 24 * time.Hour,
+		TypeCareteamInvite:       7 * 24 * time.Hour,
+		TypePasswordReset:        7 * 24 * time.Hour,
+		TypeSignUp:               31 * 24 * time.Hour,
+		TypePatientPasswordReset: 24 * time.Hour,
```
Comment (ydrisR) [inline models/confirmation.go]: Jira story indicates a timeout of 1hour, not 24...

**Review by ydrisR — CHANGES_REQUESTED** [review]

File: api/forgot.go (hunk 1 above)
Comment (franck-fourel) [reply api/forgot.go]: that's correct

File: templates/locales/fr.yaml (hunk 2 above)
Comment (franck-fourel) [reply templates/locales/fr.yaml]: ok

File: api/hydrophoneApi.go (hunk 3 above)
Comment (franck-fourel) [reply api/hydrophoneApi.go]: the new route can be removed if we are ok to consider that the same route will accept 2 different types of behavior based on the payload.

File: clients/mockStoreClient.go (hunk 5 above)
Comment (franck-fourel) [reply clients/mockStoreClient.go]: three days ... 1 hour ... let me just update the comments

File: models/confirmation.go [hunk 8]
```diff
@@ -6,6 +6,7 @@ import (
 	"encoding/json"
 	"fmt"
 	"log"
+	mathRand "math/rand"
```
Comment (yvon-dblg) [inline models/confirmation.go]: I think it is preferable to use the crypto random generator for this:
https://golang.org/pkg/crypto/rand/

Comment (franck-fourel) [reply models/confirmation.go]: that's correct.
let me push an update with crypto/rand

File: clients/mockStoreClient.go [hunk 9]
```diff
@@ -30,6 +30,12 @@ func (d *MockStoreClient) UpsertConfirmation(notification *models.Confirmation)
 	if d.doBad {
 		return errors.New("UpsertConfirmation failure")
 	}
+	if notification.Email == "patient@myemail.com" && notification.Key != "" && notification.ShortKey != "" {
```
Comment (ydrisR) [inline clients/mockStoreClient.go:39]: I don't understand the logic. This mock will always return nil...?

Comment (franck-fourel) [reply clients/mockStoreClient.go:39]: if it does not return nil, it means that key and shortKey haven't been generated; That's the objective of the test: validate both keys are generated for a patient call.

**Review by ydrisR — APPROVED** [review]
LGTM

**Pull request merged**